source/simulator/helper_functions.cc

```diff
@@ -251,6 +251,80 @@ namespace aspect
   }
 
 
+  template <int dim>
+  bool Simulator<dim>::maybe_generate_statistics(const bool statistics_already_generated)
+  {
+    // if we do not do more than one nonlinear iteration, or the
+    // cooresponding columns have already been generated in a previous call
+    // to this function, we fill the columns with their correct value
+    const bool generate_statistics_in_first_call =
+      (parameters.nonlinear_solver == NonlinearSolver::IMPES
+       || parameters.nonlinear_solver == NonlinearSolver::Advection_only
+       || parameters.max_nonlinear_iterations == 1
+       || parameters.max_nonlinear_iterations_in_prerefinement == 1);
+
+    // in this special case, we have initialized these lines already and do not
+    // want to write any more statistics for the first time step
+    if (statistics_already_generated && timestep_number == 0
+        && pre_refinement_step == 0
+        && generate_statistics_in_first_call)
+      {
+        return true;
+      }
+    // if not, generate statistics
+    else if (generate_statistics_in_first_call || statistics_already_generated)
+      {
+        // set global statistics about this time step
+        statistics.add_value("Time step number", timestep_number);
+        if (parameters.convert_to_years == true)
+          statistics.add_value("Time (years)", time / year_in_seconds);
+        else
+          statistics.add_value("Time (seconds)", time);
+
+        if (parameters.convert_to_years == true)
+          statistics.add_value("Time step size (years)", time_step / year_in_seconds);
+        else
+          statistics.add_value("Time step size (seconds)", time_step);
+
+        statistics.add_value("Number of mesh cells",
+                             triangulation.n_global_active_cells());
+
+        unsigned int n_stokes_dofs = introspection.system_dofs_per_block[0];
```

Comment:
could you replace `[0]` by `[introspection.block_indices.velocity]` here, or would that yield a different result?